[12.0][WIP] add test on all module

sebastienbeau opened the pull request:
Include PR : https://github.com/acsone/odoo-shopinvader-payment/pull/7
- [x]  test on shopinvader_payment
- [x] test on invader_payment_sips

sebastienbeau committed:
[MIG] migrate module shopinvader_quotation_payment

Migrate the module shopinvader_quotation_payment
Add module test_shopinvader_payment
Improve code and fix pep8

sebastienbeau committed:
[IMP] refactor and add test

Fix: add _invader_restrict_payment_mode_ids to be sure to not allow payment
mode not cofigured

Rename payment_mode into payment_mode_id as it's an ID

Fix: inheriting "invader.payment.service", shopinvader_payment module
should inherit with a new "_name" to avoid changing the behaviours of
base components.

Use dispatch in test instead of calling the method directly

lmignon commented on invader_payment/services/invader_payment_service.py:
```diff
@@ -34,14 +35,28 @@ def _invader_find_payable_from_transaction(self, transaction):
-        return {"target": {"type": "string", "required": True, "allowed": []}}
+        validator = {
+            "target": {"type": "string", "required": True, "allowed": []},
+            "payment_mode_id": {
```
Why????? The target must only be used to to find the payable!

sebastienbeau replied:
Indeed your right !

sebastienbeau replied:
Indeed your right, I should note inject the payment_mode_id here.
I should only add this in the payment service

lmignon commented on invader_payment/services/invader_payment_service.py:
```diff
@@ -34,14 +35,28 @@ def _invader_find_payable_from_transaction(self, transaction):
         """
         raise NotImplementedError()
 
+    def _invader_restrict_payment_mode_ids(self):
```
we don't need it

sebastienbeau replied:
Indeed I should only have this in shopinvader_payment module.
I will refactor it

lmignon replied:
but why?

sebastienbeau replied:
But we need to define which payment mode is allowed, right now you can paid with a payment mode that i snot configured on shopinvader backend

lmignon replied:
the available payment_mode ids could depend on a lot of parameters. In some cases the available payment_mode will depend on the sale order itself. (mode if total € > ....).

lmignon replied:
@sebastienbeau The validation should be at 2 levels:
Into the payment service implmentation we should check that the payment_mode_id received by each payment service is supporetd by the service.
When the payment_mode_id is applied to the payable, the payable could validate if this payment_mode_id can be applied. For the second point I would avoid to declare this check into invader.payable

sebastienbeau replied:
So the idea will to have (with stripe example)
- a check in "def confirm_payment" that check that the payment mode selected have stripe as provider.
- a check in model/sale.py of shopinvader_payment  that check that the payment_mode selected is compatible with the mode allowed on the backend? we add this in the "_invader_prepare_payment_transaction_data" ?

sebastienbeau replied:
Regarding making payment_mode_id requirer, we add this in every payment service (stripe, manual, sips), do we copy this or do we try to share it between all service?

lmignon replied:
> So the idea will to have (with stripe example)
> 
>     * a check in "def confirm_payment" that check that the payment mode selected have stripe as provider.
> 
:+1: 

>     * a check in model/sale.py of shopinvader_payment  that check that the payment_mode selected is compatible with the mode allowed on the backend? we add this in the "_invader_prepare_payment_transaction_data" ?

:+1:

lmignon replied:
> Regarding making payment_mode_id requirer, we add this in every payment service (stripe, manual, sips), do we copy this or do we try to share it between all service?

I prefer to keep the api without dependencies.... -> copy

sebastienbeau replied:
Ok I am going to do this I push it.
thanks for your usefull comment

lmignon reviewed (changes requested):
@sebastienbeau I don't understand at all whyyou modify the target validator....

sebastienbeau committed:
[REF] implement check on allowed payment on payable

sebastienbeau committed:
[IMP] Improve test and check acquirer

Check that acquirer match the service with test
Add test on manual payment

sebastienbeau committed:
[REF] start to test payment_sips

Restore payment_mode_id validator
Start adding test
Add demo data

sebastienbeau committed:
[REF] make the code consistent

lmignon commented on shopinvader_payment/services/abstract_payable_sale.py:
```diff
@@ -10,7 +10,7 @@ class AbstractPayableSaleService(AbstractComponent):
     _name = "shopinvader.abstract.payable.sale.service"
 
-    def _get_available_payment_mode(self, sale):
+    def _get_available_payment_method(self, sale):
```
you should use plurals... `s/_get_available_payment_method/_get_available_payment_methods/`

lmignon commented on invader_payment_stripe/services/payment_stripe.py:
```diff
@@ -120,32 +124,36 @@ def confirm_payment(self, target, **params):
+
+        acquirer = payment_mode.payment_acquirer_id.sudo()
+        if acquirer.provider != "stripe":
+            raise UserError(_("Payment mode acquirer mismatch."))
```
@sebastienbeau The error message doesn't explain the problem... The issue here is that the stripe service is called with a payment_mode  linked to a non stripe acquirer... (xyz != stripe)

sebastienbeau replied:
Payment mode acquirer mismatch => Payment mode acquirer mismatch, should be 'stripe' found 'transfer' ?

sebastienbeau replied:
Payment mode acquirer mismatch, should be 'stripe' instead of 'transfer'

lmignon commented on test_shopinvader_payment/tests/test_payment.py:
```diff
@@ -0,0 +1,53 @@
+# -*- coding: utf-8 -*-
```
@sebastienbeau This file is a new one... The copyright year should be 2019. The authors should put into the readme.The coding can be removed.
This test file is used to test the shopinvader manuel payment.. Its name should be `test_shopinvader_payment_manual` and the classname `TestShopinvaderPaymentManual()`

lmignon commented on test_shopinvader_payment/tests/test_quotation.py:
```diff
@@ -22,40 +21,47 @@ def test_get_no_cart_payment_info(self):
 class ShopinvaderQuotationPaymentCase(CommonConnectedQuotationCase):
```
here you test the shopinvader_quotation_payment addon. Therefore, the file name should be `test_shopinvader_qutation_payment.py` and the classname `TestShopinvaderQuotationPayment`

lmignon commented on test_shopinvader_payment/tests/test_quotation.py:
```diff
@@ -22,40 +21,47 @@ def test_get_no_cart_payment_info(self):
+        with self.work_on_services(
+            partner=self.partner, shopinvader_session=self.shopinvader_session
+        ) as work:
+            self.payment_service = work.component(usage="payment_manual")
```
s/payment_service/payment_manual_service/

lmignon commented on test_shopinvader_payment/tests/test_quotation.py:
```diff
@@ -22,40 +21,47 @@ def test_get_no_cart_payment_info(self):
+        ) as work:
+            self.payment_service = work.component(usage="payment_manual")
 
     def test_get_cart_payment_info(self):
```
It's difficult to understand the purpose of the test. We should improve the description by describing the data, the test case, the expected result

```python
def test_get_cart_payment_info(self):
    """
    Data:
        A sale order with a product requiring a quotation
    Test Case:
        Retrieve the quotation info
    Expected result:
        'available_methods' must be present into the informations
    """

lmignon commented on test_shopinvader_payment/tests/test_quotation.py:
```diff
@@ -22,40 +21,47 @@ def test_get_no_cart_payment_info(self):
-        self.cart.order_line[0].product_id.only_quotation = True
-        response = self.service.dispatch("search", params={"id": self.cart.id})
+        self.quotation.order_line[0].product_id.only_quotation = True
+        response = self.service.dispatch(
```
You don't test the quoration service... Am I wrong?

lmignon replied:
self.service is the cart service

sebastienbeau replied:
self.service here is quotations service https://github.com/shopinvader/odoo-shopinvader/blob/12.0/shopinvader_quotation/tests/test_quotation.py#L40
Maybe we should put the name explicitly

lmignon replied:
@sebastienbeau I miss something... what's the purpose of the quotation service? When do you use the quotation service or the car service.I can't get a clear conceptual representation between cart and quotation service

sebastienbeau replied:
When you have a cart you can pay it or request a quotation. (And in the case that you have a product that required quotation, you can not pay it).
When requesting a quotation on the cart https://github.com/shopinvader/odoo-shopinvader/blob/12.0/shopinvader_quotation/services/cart.py#L12. The typology of the cart change and it's set to "quotation".

So now this is not a cart anymore, it's a quotation in the state "estimating" https://github.com/shopinvader/odoo-shopinvader/blob/12.0/shopinvader_quotation/models/sale.py#L13

Odoo user will update the price of the quotation and click on "sent by email". state => estimated

Now the quotation can be accept and paid by the user using a payment method

sebastienbeau replied:
@lmignon tell me if you need more info

sebastienbeau replied:
I am going to clean improve the test. I reuse existing from 10. But method name are not really explicit

sebastienbeau committed:
[IMP] improve code, add missing readme

lmignon commented on test_shopinvader_payment/tests/test_shopinvader_quotation_payment.py:
```diff
@@ -0,0 +1,107 @@
+        Test Case:
+            Retrieve the quotation info
+        Expected result:
+            'available_methods' must not have all payment mode
```
@sebastienbeau I don't understant the difference between this test and the previous one. Both expect the same result...

sebastienbeau replied:
You have product that are flagged "only_quotation" and this remote all available payment method on the cart.
This limitation should only impact cart service and never the quotation service.
this is why I test the two case of product. Only to be sure that this retriction do not impact quotation service.

lmignon replied:
You says into the expected result that no all the response should not holds all the payment mode but you check the contrary

sebastienbeau replied:
ououlala, I need to change my glass ;)

lmignon commented on test_invader_payment/tests/test_stripe.py:
```diff
@@ -65,9 +72,31 @@ def test_confirm_payment_two_step(self):
         self.assertEqual(result, {"success": True})
+
+    def test_wrong_provider_confirm(self):
+        self.payment_mode = self.env.ref(
```
@sebastienbeau `payment_mode`should be scoped to the method not to the instance...
`s/self.payment_mode/payment_mode_check/`

lmignon commented on test_invader_payment/tests/test_sips.py:
```diff
@@ -0,0 +1,86 @@
+        self.assertEqual(response.status_code, 200)
+
+    def test_wrong_provider_prepare_payment(self):
+        self.payment_mode = self.env.ref(
```
@sebastienbeau payment_modeshould be scoped to the method not to the instance...
s/self.payment_mode/payment_mode_check/

lmignon commented on test_invader_payment/tests/test_manual.py:
```diff
@@ -0,0 +1,50 @@
+        self.assertEqual(transaction.state, "pending")
+
+    def test_wrong_provider_add_payment(self):
+        self.payment_mode = self.env.ref(
```
@sebastienbeau payment_modeshould be scoped to the method not to the instance...
s/self.payment_mode/payment_mode_check/

lmignon commented on shopinvader_quotation_payment/services/invader_payment_service.py:
```diff
@@ -0,0 +1,30 @@
+            if not quotation:
+                raise UserError(_("The quotation do not exist"))
+            elif quotation.state != "sent":
+                raise UserError(_("The quotation have been not yet estimated"))
```
```suggestion
                raise UserError(_("The quotation is not yet estimated"))
```

lmignon commented on shopinvader_quotation_payment/services/invader_payment_service.py:
```diff
@@ -0,0 +1,30 @@
+                raise UserError(_("Quotation id is missing"))
+            quotation = self.component(usage="quotations")._get(quotation_id)
+            if not quotation:
+                raise UserError(_("The quotation do not exist"))
```
```suggestion
                raise UserError(_("The quotation doesn't exist."))
```

lmignon commented on shopinvader_quotation_payment/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,5 @@
+
+Estimated quotation can be paid with all payment mode configured on shopinvader backend.
+
+Cart that containt a product that required a quotation can not be paid anymore (no payment mode available). The request_quotation must be called to convert the cart.
```
```suggestion
Cart that contains a product that requires a quotation can not be paid anymore (no payment mode available). The method `request_quotation` must be called on the cart service to convert the quotation into a cart.
```

sebastienbeau committed:
[REF] simplify code

add _check_acquirer method in payment_service
add a property for get invader.payment component
